### PR TITLE
Add cases for subcommand convert of qemu-img

### PR DESCRIPTION
**Pull request by pingl2016**
1. Add cases for cluster size/sparse size/cache mode/show progress
2. Add function verify_info() to check result of convert operation

Signed-off-by: Ping Li pingl@redhat.com

### CODE REVIEW

**Comment by pingl2016** [comment]
ID: 1360544, 1361861, 1362009, 1362614

**Comment by pingl2016** [comment]
avocado-framework/avocado-vt#649

File: qemu/tests/qemu_img.py [hunk 1]
```diff
@@ -177,6 +178,45 @@ def create_test(cmd):
                 cluster_size=params.get("image_cluster_size"))
         remove(img)
 
+    def send_signal(timeout=360):
```
Comment (vivianQizhu) [inline qemu/tests/qemu_img.py]: Suggest to set timeout value in config file.

Comment (pingl2016) [inline qemu/tests/qemu_img.py]: For this case, the default value is enough. Keep the parameter here is for the subsequent cases we develop.

**Comment by vivianQizhu** [comment]
Not sure if it was relevant with this patch, but I hit an error for all cases in qemu_disk_img.convert:
12:04:57 ERROR|
12:04:57 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py:420
12:04:57 ERROR| Traceback (most recent call last):
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
12:04:57 ERROR|     self._runTest()
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
12:04:57 ERROR|     run_func(self, params, env)
12:04:57 ERROR|   File "/home/tp-qemu/qemu/tests/qemu_disk_img_convert.py", line 53, in run
12:04:57 ERROR|     convert_test.start_vm(n_params)
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
12:04:57 ERROR|     return fn(*args, **kwargs)
12:04:57 ERROR|   File "/home/tp-qemu/qemu/tests/qemu_disk_img.py", line 68, in start_vm
12:04:57 ERROR|     vm.wait_for_login(timeout=login_timeout)
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/virt_vm.py", line 1040, in wait_for_login
12:04:57 ERROR|     s_session = self.serial_login(username=username, password=password)
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
12:04:57 ERROR|     return fn(*args, **kwargs)
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/virt_vm.py", line 1199, in serial_login
12:04:57 ERROR|     remote.handle_prompts(console, username, password, prompt, timeout)
12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/virttest/remote.py", line 172, in handle_prompts
12:04:57 ERROR|     raise LoginAuthenticationError(msg, text)
12:04:57 ERROR| LoginAuthenticationError: Got username prompt twice    (output: '\nLogin incorrect\n\ndhcp-9-53 login: ')
12:04:57 ERROR|
12:04:57 ERROR| Traceback (most recent call last):

12:04:57 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-39.0-py2.7.egg/avocado_vt/test.py", line 434, in _run_avocado
    raise test_exception

12:04:57 ERROR| LoginAuthenticationError: Got username prompt twice    (output: '\nLogin incorrect\n\ndhcp-9-53 login: ')

12:04:57 ERROR| ERROR 02-smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.my-provider.qemu_disk_img.convert.base_to_qcow2.compressed -> LoginAuthenticationError: Got username prompt twice    (output: '\nLogin incorrect\n\ndhcp-9-53 login: ')

File: qemu/tests/qemu_img.py [hunk 2]
```diff
@@ -177,6 +178,45 @@ def create_test(cmd):
...
+        to report progress
+        """
+        logging.info("Send signal to qemu-img")
+        end_time = time.time() + timeout
+        while time.time() < end_time:
+            time.sleep(1)
+            status = utils.system("kill -SIGUSR1 `pidof qemu-img`",
+                                  ignore_status=True)
+            if status == 0:
+                return
```
Comment (xutian) [inline qemu/tests/qemu_img.py]: better to  return None

File: qemu/tests/qemu_img.py [hunk 3]
```diff
@@ -177,6 +178,45 @@ def create_test(cmd):
...
+    def check_command_output(CmdResult):
+        """
+        Check standard error or standard output of command
+        : param CmdResult: a list of CmdResult objects
+        """
+        logging.info("Check result of command")
+        check_output = params.get("check_output", "")
+        if not check_output:
+            if CmdResult.exit_status == 0:
+                return
```
Comment (xutian) [inline qemu/tests/qemu_img.py]: same

File: qemu/tests/qemu_img.py [hunk 4]
```diff
@@ -177,6 +178,45 @@ def create_test(cmd):
...
+            else:
+                err_msg = "Get nonzero exit status '%s'" % CmdResult.command
+                raise error.TestFail(err_msg)
+        if check_output == "stderr":
```
Comment (xutian) [inline qemu/tests/qemu_img.py]: how about   
if not hasattr(CmdResult, check_output):
     output = ""
else:
     output = getattr(CmdResult, check_output)

File: qemu/tests/qemu_img.py [hunk 5]
```diff
@@ -4,6 +4,7 @@
 import commands
 import shutil
 import tempfile
+import time
```
Comment (xutian) [inline qemu/tests/qemu_img.py]: module import sequence should be
```
import os
import time
import re
....
```

File: qemu/tests/qemu_disk_img.py [hunk 6]
```diff
@@ -130,6 +132,39 @@ def get_info(self):
...
     @error.context_aware
+    def verify_info(self, params=None):
+        """
+        verify option is applied to image file correctly
+        """
+        error.context("verify option of converted image", logging.info)
+        image_filename = storage.get_image_filename(params, self.data_dir)
+        info = utils_test.get_image_info(image_filename)
+        avalue = evalue = ""
+        option_verified = params.get("option_verified", "").split()
```
Comment (xutian) [inline qemu/tests/qemu_disk_img.py]:
```
for option in params.objects("option_verified"):
     avalue = xxx
```

**Comment by pingl2016** [comment]
@xutian patch has been updated, could you help to review it again? thanks

File: qemu/tests/qemu_img.py [hunk 7]
```diff
@@ -177,6 +178,42 @@ def create_test(cmd):
...
+        logging.info("Fail to get pid of qemu-img")
+
+    def check_command_output(CmdResult):
+        """
+        Check standard error or standard output of command
+        : param CmdResult: a list of CmdResult objects
+        """
+        logging.info("Check result of command")
+        check_output = params.get("check_output", "exit_status")
+        output = getattr(CmdResult, check_output)
```
Comment (xutian) [inline qemu/tests/qemu_img.py]: exception maybe raised here, better to test with hasattr then getattr.

Comment (xutian) [inline qemu/tests/qemu_img.py]: it looks logical should be like below, right?

```
if not hasattr(CmdResult, check_output):
      raise exceptions.TestError("Unknow check output '%s'" % check_output)
output = getattr(CmdResult, check_output)
if check_output == "exit_status" and output == 0:
     return None
if check_output == "exit_status" and output != 0:
     raise exceptions.TestFail("cmd exit with nonzero status: %s" % CmdResult.command)
pattern = xxx
if not re.findall(pattern, output):
     err_msg = "xxxx"
     raise exceptions.TestFail(err_msg)
```

Comment (pingl2016) [inline qemu/tests/qemu_img.py]: I think it' necessary to check check_output, as it is added in configure file by developer when add a new case. If the issue happen, I think it maybe a bug introduced by the case developer.

**Comment by pingl2016** [comment]
Results of case qemu_disk_img:
RESULTS    : PASS 51 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
Results of case qemu_img:
RESULTS    : PASS 31 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0

(14/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.compat.compat_0.10.cluster_size.cluster_size_512:  PASS (435.23 s)
 (15/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.compat.compat_0.10.cluster_size.cluster_size_2k:  PASS (214.59 s)
 (16/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.compat.compat_0.10.cluster_size.cluster_size_2M:  PASS (203.25 s)
 (20/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.compat.compat_1.1.cluster_size.cluster_size_512:  PASS (419.88 s)
 (21/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.compat.compat_1.1.cluster_size.cluster_size_2k:  PASS (211.08 s)
 (22/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.compat.compat_1.1.cluster_size.cluster_size_2M:  PASS (210.78 s)
(23/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.fully_allocated:  PASS (370.65 s)
 (24/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.cache_mode.none:  PASS (233.29 s)
 (25/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.cache_mode.wirteback:  PASS (198.98 s)
 (26/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.cache_mode.unsafe:  PASS (203.78 s)
 (27/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.cache_mode.directsync:  PASS (1275.89 s)
 (28/51) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img.convert.base_to_qcow2.cache_mode.writethrough:  PASS (931.19 s)

 (18/31) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_img.convert.to_qcow2.show_progress.on:  PASS (86.61 s)
 (19/31) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_img.convert.to_qcow2.show_progress.off:  PASS (443.35 s)

File: qemu/tests/qemu_img.py [hunk 8]
```diff
@@ -177,6 +178,42 @@ def create_test(cmd):
...
+        if check_output == "exit_status":
+            if output == 0:
+                return None
+            else:
+                err_msg = "Get nonzero exit status '%s'" % CmdResult.command
```
Comment (PandaWei) [inline qemu/tests/qemu_img.py]: It's better to output the value of output.
"Get nonzero exit status(%d) '%s'" % (output, CmdResult.command)

File: qemu/tests/qemu_img.py [hunk 9]
```diff
@@ -177,6 +178,42 @@ def create_test(cmd):
...
+                raise error.TestFail(err_msg)
+        pattern = params.get("command_result_pattern")
+        if not re.findall(pattern, output):
+            err_msg = "Fail to get expected result!"
+            err_msg += "the output is: '%s', " % output
+            err_msg += "the expected pattern is: '%s'" % pattern
```
Comment (PandaWei) [inline qemu/tests/qemu_img.py]: Here simple words looks better.
I like the format you used in #731 
 "Expect: %s, actual: %s" % (expect, actual)

How about "Output: %s, Expected pattern: %s" % (output, pattern)

File: qemu/tests/qemu_img.py (hunk 7 above)
Comment (PandaWei) [inline qemu/tests/qemu_img.py]: And I agree with @xutian ‘s suggestion.
Do check by hasattr() firstly is not bad.

**Comment by PandaWei** [comment]
Other code looks good.

File: qemu/tests/qemu_img.py (hunk 9 above)
Comment (pingl2016) [inline qemu/tests/qemu_img.py]: yes, will update

File: qemu/tests/qemu_img.py (hunk 7 above)
Comment (pingl2016) [inline qemu/tests/qemu_img.py]: The exception only raised if class CmdResult changed or configure file is wrong. The error cannot happen in the release code. If we add the code, it is only used for debugging case. So I still think it is unnecessary to check. @PandaWei

File: qemu/tests/qemu_img.py (hunk 8 above)
Comment (pingl2016) [inline qemu/tests/qemu_img.py]: will add it

**Comment by pingl2016** [comment]
@xutian @PandaWei patch updated, could you help to review it again?

File: qemu/tests/cfg/qemu_disk_img.cfg [hunk 10]
```diff
@@ -60,26 +62,66 @@
                                     cluster_size = 2M
                         - compat:
                             no Host_RHEL.m5, Host_RHEL.m6
+                            option_verified += " compat"
                             variants:
```
Comment (suqinhuang) [inline qemu/tests/cfg/qemu_disk_img.cfg]: why not 
variants:
- compat_0.10:
- compat_1.1:
- variants:
 - default:
....

File: qemu/tests/cfg/qemu_img.cfg [hunk 11]
```diff
@@ -64,6 +64,17 @@
                             #cluster_size set cluster size used in qemu-img -o cluster_size=, default is 64k.
                             cluster_size = 2048
                             no Host_RHEL.m5
+                        - show_progress:
+                            command_result_pattern = "\(\d+\.\d+/100%\)"
+                            variants:
+                                - on:
```
Comment (suqinhuang) [inline qemu/tests/cfg/qemu_img.cfg]: on/off is a general name, it's maybe duplicated in other test, it's better to set a special name.

File: qemu/tests/qemu_img.py [hunk 12]
```diff
@@ -195,6 +232,9 @@ def _convert(cmd, output_fmt, img_name, output_filename,
             cmd += " -c"
         if encrypted == "yes":
             cmd += " -e"
+        show_progress = params.get("show_progress", "")
```
Comment (suqinhuang) [inline qemu/tests/qemu_img.py]: what's the default value of show_progress

Comment (pingl2016) [inline qemu/tests/qemu_img.py]: an empty string

**Comment by suqinhuang** [comment]
Look good to me

**Pull request merged**